### PR TITLE
Record fix -- LogReaderSpec and SimpleLogSpec (LogWriterSpec)

### PR DESCRIPTION
**Pull request by saiseekerhn**

### CODE REVIEW

File: disk/test/com/treode/disk/edit/SimpleLogSpec.scala [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+   def record (s: String): Async [Unit] =
+   async { cb =>
+      // read pos is aligned at start, and after each record
+      // write pos is at end of last record
+      if (buf.writePos > 0) {
+         buf.writePos -= 1
+         buf.writeByte(1)
+      }
+      buf.writeString (s)
+      buf.writeByte(0)
```
Comment (Topher-the-Geek) [inline disk/test/com/treode/disk/edit/SimpleLogSpec.scala]: This way, `end` is one byte after the end of the last record. If you move buf.writeByte(0) in between line 34 and 35, then `end` will be the end of the last record. Then you won't need to conditionally subtract 1 from writePos above.

File: disk/test/com/treode/disk/edit/SimpleLogSpec.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      implicit val scheduler = StubScheduler.random()
+      var testfile = StubFile(new Array[Byte](50), 0)
+      val rec = new SimpleLog(testfile, DriveGeometry(10, 10, 16384))
+      var str = "hithere"
+      rec.record(str).expectPass()
+      
+      var input = PagedBuffer(12)
+      testfile.fill (input, 0, str.length()) .expectPass()
+      
+      var actual = input.readString()
```
Comment (Topher-the-Geek) [inline disk/test/com/treode/disk/edit/SimpleLogSpec.scala]: This is strange:

    var x = in.readString()
    x += in.readByte()
    assert (x == "string"+0)

You can just say:

    assert (in.readString() == "string")
    assert (in.readByte() == 0)

File: disk/test/com/treode/disk/edit/SimpleLogSpec.scala [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+   import com.treode.disk.DriveGeometry
+   
+   "SimpleLog" should "record once to a StubFile" in {
+      implicit val scheduler = StubScheduler.random()
+      var testfile = StubFile(new Array[Byte](50), 0)
+      val rec = new SimpleLog(testfile, DriveGeometry(10, 10, 16384))
+      var str = "hithere"
+      rec.record(str).expectPass()
+      
+      var input = PagedBuffer(12)
```
Comment (Topher-the-Geek) [inline disk/test/com/treode/disk/edit/SimpleLogSpec.scala]: You need to deal with PagedBuffer in the production code. In the test code, you can take a shortcut:

    val input = new ArrayBuffer (testfile.data)

Then you can skip the fill business.

File: disk/test/com/treode/disk/edit/LogReaderSpec.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import com.treode.async.stubs.implicits._
+import com.treode.async.Globals
+
+class LogReader (file: StubFile, geom: DriveGeometry) {
+   
+   var buf = PagedBuffer(12)
+   var pos = 0
+   var end = 1
+   
+   def read(): Async[Seq[String]] = {
```
Comment (Topher-the-Geek) [inline disk/test/com/treode/disk/edit/LogReaderSpec.scala]: This requires the user to call read repeatedly, and each time it yields a sequence of length 1. We want to allow the user to call read once, and this returns a sequence of all the strings written.

**Comment by saiseekerhn** [comment]
when you say testfile.data -- wouldn't I have to fill that with my string still?

**Comment by saiseekerhn** [comment]
would I do 
val input = new ArrayBuffer (testfile.data)
input.writeString(str)

testfile.data would consist of the str already right?

**Comment by Topher-the-Geek** [comment]
In the case of LogReaderSpec, you're doing:

```
var input = PagedBuffer(12)
input.writeString (str)
input.writeByte(0)
testfile.flush (input, 0).expectPass()
```

You can just do

```
val buf = new ArrayBuffer (testfile.data)
buf.writeString (str)
buf.writeByte(0)
```

A similar shortcut can be taken in SimpleLogSpec for reading items from the test file.

**Comment by saiseekerhn** [comment]
So the shortcut worked for SimpleLogSpec but for LogReaderSpec

I tried:

var testfile = StubFile(new Array[Byte](50), 0)
var buf = ArrayBuffer(testfile.data)
buf.writeString (str)   // this line is an error: Buffer overflow, 1 required, 0 available
buf.writeByte(0)
var logreader = new LogReader(testfile, DriveGeometry(10, 10, 16384))

**Comment by saiseekerhn** [comment]
The buffer overflow seems to occur because there is not any available space to write. Comes from the method at com.treode.buffer.ArrayBuffer.requireWritable(ArrayBuffer.scala:61)

**Comment by Topher-the-Geek** [comment]
Yeah, the ArrayBuffer does not grow space, so you need to give yourself some space first.

```
var testfile = StubFile(1 << 16, 0)
var buf = ArrayBuffer(testfile.data)
buf.writeString (str) // no more error
buf.writeByte(0)
```

**Comment by saiseekerhn** [comment]
hmm...the error still shows up

**Comment by Topher-the-Geek** [comment]
I've changed the constructor methods for ArrayBuffer. Now there's [readable and writable](https://github.com/Treode/store/commit/b649e363707c90078a820ca4d580a2b1188fa1c1). You can use each as needed, rather than needing to fiddle with `writePos`.

**Comment by saiseekerhn** [comment]
Thanks that worked!

**Pull request closed without merging**